### PR TITLE
HSM: Support multiple signers (#539)

### PR DESCRIPTION
**Pull request by ximon18**
Separates out signer dispatch from KrillSigner to a new SignerProvider struct so that SignerProvider can handle the noisy enum based dispatch to a particular Signer implementation and KrillSigner can focus on business logic. Adds a 'hsm' feature gate and puts the a 'DummySigner' behind it.

### CODE REVIEW

File: src/commons/crypto/signing.rs [hunk 1]
```diff
@@ -36,84 +40,244 @@ use crate::{
...
+                key,
+                &base_repo.ca_repository(name_space).join(&[]).unwrap(), // force trailing slash
+                &base_repo.rpki_manifest(name_space, &pub_key.key_identifier()),
+                Some(&base_repo.rpki_notify()),
+            )
+            .map_err(crypto::Error::signing)
+        }
+
+        let enc = match self {
+            SignerProvider::OpenSsl(signer) => func(signer.deref(), base_repo, name_space, key),
```
Comment (timbru) [inline src/commons/crypto/signing.rs:133]: I am (generally) not fond of fn names like 'func'. Perhaps a more descriptive thing like 'sing_csr_with_signer'? Or.. is there a way to use the match to get the signer to use without the need for an inner fn?

File: src/commons/util/dummysigner.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
+use rpki::repository::crypto::{KeyIdentifier, PublicKey, PublicKeyFormat, Signature, SignatureAlgorithm, Signer};
+
+use super::softsigner::SignerError;
+
+/// A dummy signer to prove that compilation with two different Signer implementations works
```
Comment (timbru) [inline src/commons/util/dummysigner.rs:5]: That makes sense for now, but will this be removed when real signer implementations are added?

**Review by timbru — APPROVED** [review]
Looks like a good architecture. I had to minor remarks - but neither are show stoppers at this time.

File: src/commons/crypto/signing.rs (hunk 1 above)
Comment (ximon18) [reply src/commons/crypto/signing.rs:133]: Oops, that should indeed not be called `func`.

File: src/commons/util/dummysigner.rs (hunk 2 above)
Comment (ximon18) [reply src/commons/util/dummysigner.rs:5]: Indeed. The other PR I am working (https://github.com/NLnetLabs/krill/pull/679) on already removes it (well, replaces it with a real KMIP signer).

**Comment by ximon18** [comment]
Discussed with @timbru that we will not merge this until after the coming v0.9.2 release just to keep our lives that tiny bit easier.

**Pull request merged**